services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/PushStatusCollector.java

```diff
@@ -0,0 +1,231 @@
+    this.daVinciPushStatusScanPeriodInSeconds = daVinciPushStatusScanIntervalInSeconds;
+    this.pushStatusStoreScanExecutor = Executors.newFixedThreadPool(pushStatusScanThreadNumber);
+  }
+
+  public void start() {
+    if (daVinciPushStatusScanEnabled) {
+      offlinePushCheckScheduler
+          .scheduleAtFixedRate(this::scanDaVinciPushStatus, 0, daVinciPushStatusScanPeriodInSeconds, TimeUnit.SECONDS);
+    } else {
+      LOGGER.warn("Offline push monitoring Da Vinci push status is not enabled, will only check server push status.");
+    }
+  }
+
+  public void subscribeTopic(String topicName, int partitionCount) {
+    String storeName = Version.parseStoreFromKafkaTopicName(topicName);
+    Store store = storeRepository.getStore(storeName);
+    if (store == null) {
+      LOGGER.warn("Store {} not found in store repository, will not monitor Da Vinci push status", storeName);
+    } else {
+      if (daVinciPushStatusScanEnabled && store.isDaVinciPushStatusStoreEnabled()
+          && Version.parseVersionFromKafkaTopicName(topicName) > 1) {
+        LOGGER.info("Will monitor Da Vinci push status for topic {}", topicName);
+        topicToPushStatusMap.put(topicName, new TopicPushStatus(topicName, partitionCount));
+      }
+    }
+  }
+
+  public void unsubscribeTopic(String topicName) {
+    topicToPushStatusMap.remove(topicName);
+  }
+
+  private void scanDaVinciPushStatus() {
+    List<CompletableFuture<TopicPushStatus>> resultList = new ArrayList<>();
+    for (Map.Entry<String, TopicPushStatus> entry: topicToPushStatusMap.entrySet()) {
+      String topicName = entry.getKey();
+      TopicPushStatus pushStatus = entry.getValue();
+      if (!pushStatus.isMonitoring()) {
+        continue;
+      }
+      if (pushStatus.getDaVinciStatus() != null && pushStatus.getDaVinciStatus().getStatus().isTerminal()) {
+        resultList.add(CompletableFuture.completedFuture(pushStatus));
+      } else {
+        resultList.add(CompletableFuture.supplyAsync(() -> {
+          ExecutionStatusWithDetails statusWithDetails = PushMonitorUtils.getDaVinciPushStatusAndDetails(
+              pushStatusStoreReader,
+              topicName,
+              pushStatus.getPartitionCount(),
+              Optional.empty());
+          pushStatus.setDaVinciStatus(statusWithDetails);
+          return pushStatus;
+        }, pushStatusStoreScanExecutor));
+      }
+    }
+    // Collect the executor result and compute aggregate results for ongoing pushes.
+    for (CompletableFuture<TopicPushStatus> future: resultList) {
+      TopicPushStatus pushStatus;
+      try {
+        pushStatus = future.get();
+      } catch (Exception e) {
+        LOGGER.error("Caught exception when getting future result of push status.", e);
+        continue;
+      }
+      ExecutionStatusWithDetails daVinciStatus = pushStatus.getDaVinciStatus();
+      ExecutionStatusWithDetails serverStatus = pushStatus.getServerStatus();
+      if (serverStatus == null) {
```

Comment:
curious, when this can happen?

Reply:
Do you mean serverStatus == null?
It will happen when DVC status scanner is scanning a topic but no server terminal status is reported yet.